### DISCUSSION TITLE
Question on Generic Param Binding Changes 1.1.354

### DISCUSSION THREAD

**Discussion by max-muoto** [microsoft/pyright#8004] [General]
It seems in 1.1.354, the behavior for something like changed:

```python
"""Manage jobs that can be dispatched to a worker pool."""

import dataclasses
from typing import TYPE_CHECKING, Any, Callable, Generic, TypeVar

from collections.abc import Awaitable

T = TypeVar("T")


@dataclasses.dataclass
class _BaseJob(Generic[T]):
    """Job that can be dispatched to a worker pool."""

    fn_name: str
    args: tuple[Any, ...]
    kwargs: dict[str, Any]
    queue: str
    return_type: T
    _try: int = 0

    @classmethod
    def from_callable(
        cls,
        fn: Callable[..., T],
        *args: Any,
        **kwargs: Any,
    ) -> "_BaseJob[T]":
        """Construct a job from a callable."""
        return cls(
            fn.__name__,
            args,
            kwargs,
            queue="default",
            return_type=T,
        )

    def result(self) -> T | Awaitable[T]:
        """Wait for the job to complete and return the result."""
        raise NotImplementedError


@dataclasses.dataclass
class Job(_BaseJob[T]):
    def result(self) -> T: ...

def foo(x: int) -> int: ...


job = _BaseJob.from_callable(foo, 1, 2, 3)
```

So that job when from being `job[int]` to `job[unknown]`, I presumed maybe this was because `T` wasn't actually getting used anywhere in BaseJob's init, however, even changing to this, leads to `BaseJob[Unknown]`:

```python
"""Manage jobs that can be dispatched to a worker pool."""

import dataclasses
from typing import TYPE_CHECKING, Any, Callable, Generic, TypeVar

from collections.abc import Awaitable

T = TypeVar("T")


@dataclasses.dataclass
class _BaseJob(Generic[T]):
    """Job that can be dispatched to a worker pool."""

    fn_name: str
    fn: Callable[..., T]
    args: tuple[Any, ...]
    kwargs: dict[str, Any]
    queue: str
    _try: int = 0

    @classmethod
    def from_callable(
        cls,
        fn: Callable[..., T],
        *args: Any,
        **kwargs: Any,
    ) -> "_BaseJob[T]":
        """Construct a job from a callable."""
        return cls(
            fn.__name__,
            fn,
            args,
            kwargs,
            queue="default"
        )

    def result(self) -> T | Awaitable[T]:
        """Wait for the job to complete and return the result."""
        raise NotImplementedError


@dataclasses.dataclass
class Job(_BaseJob[T]):
    def result(self) -> T: ...


def foo(x: int) -> int: ...


job = _BaseJob.from_callable(foo, 1, 2, 3)
```

[Pyright Playground
](https://pyright-play.net/?code=EQiyEMDtwcwUwAQCsD2AjAzggLgC3NggMZQJqIAmAlhgA4FG5wU4oLgIDuKATgNZweCWihQAbAHQhgAKBlUAtiJ6EKBcETHgMGOBhkAzHigU4AnrSqQYCRcsIAVAJoAFAKIB9AMIAJN14BpAEkAOQBxABoEAEFIMyivcDEtNDE4KLC4SEEqIiiHCzgANXAeOSMTYnE0omwqFEgMCXA0IlslXkJoznAqbBa0uQcEAF4EAtpi0oAKYAdgAEo5GQABNX7NbV0m9Y0tHRlNnQQPACFtOAApdGnM7J5cgG0HAF0FgC4ZBG%2BEaWu0HD4QgkSBkSg0ejYRjMVjsLi8ARCETiKQgOQ-BAGSAeaAKODvBAYbBlDFYgmJZIDOCPCS0-IvL4-UowDAE7AAV1oaUesXiCFpEgZGL4PR4LIJ1FqjyJPCivKFPwAjuy4CqCTLGd8PMSzASrIQxgAGdE-FZHDB4vAoCiahAUOAGTHGBQeEiU1Jwaa2jGaDARb0-MkICkpbkC%2Bn%2BjEYgBUzNZMTikaj32j0ZFcYJvKT3wWCAAtAA%2BX5nC7-Z4vYCfZPfaReBoy9m1OGoAEVUwcN2huCo2TVhA8OAcnig31evukyASDw48B46fZ8dYhd9uPL6vpsV%2BgPJ5WquAjYD2gzgdlibC96tLW1H-t6U-Yaa6MQGXOF8YIAA%2BMR6fSp5ar1bSAA6r0hAGLwgKIC2sJECYXKDogUAsAOQ6gngiADhg949tu-a9LoCAhCg2BBEoaR4pA2DMG4PDGGUchrOo5p6BIuzmoc%2BxYP80wlroZavB814OreWFno%2BcDPq%2BRYOASArLIxGycSxbGcRxWwxBgZiQEQ3G8Vc6DloJGLaFpbQ3ph97iZJ%2BbSbJtLLDe4EoNMAAeeqUVJtiUXZEjLNBYx6f8Ehtq6SRdtMTlRAAjFEABMUQAMwLEAA)

[1.1.153 Playground
](https://pyright-play.net/?pyrightVersion=1.1.353&code=EQiyEMDtwcwUwAQCsD2AjAzggLgC3NggMZQJqIAmAlhgA4FG5wU4oLgIDuKATgNZweCWihQAbAHQhgAKBlUAtiJ6EKBcETHgMGOBhkAzHigU4AnrSqQYCRcsIAVAJoAFAKIB9AMIAJN14BpAEkAOQBxABoEAEFIMyivcDEtNDE4KLC4SEEqIiiHCzgANXAeOSMTYnE0omwqFEgMCXA0IlslXkJoznAqbBa0uQcEAF4EAtpi0oAKYAdgAEo5GQABNX7NbV0m9Y0tHRlNnQQPACFtOAApdGnM7J5cgG0HAF0FgC4ZBG%2BEaWu0HD4QgkSBkSg0ejYRjMVjsLi8ARCETiKQgOQ-BAGSAeaAKODvBAYbBlDFYgmJZIDOCPCS0-IvL4-UowDAE7AAV1oaUesXiCFpEgZGL4PR4LIJ1FqjyJPCivKFPwAjuy4CqCTLGd8PMSzASrIQxgAGdE-FZHDB4vAoCiahAUOAGTHGBQeEiU1Jwaa2jGaDARb0-MkICkpbkC%2Bn%2BjEYgBUzNZMTikaj32j0ZFcYJvKT3wWCAAtAA%2BX5nC7-Z4vYCfZPfaReBoy9m1OGoAEVUwcN2huCo2TVhA8OAcnig31evukyASDw48B46fZ8dYhd9uPL6vpsV%2BgPJ5WquAjYD2gzgdlibC96tLW1H-t6U-Yaa6MQGXOF8YIAA%2BMR6fSp5ar1bSAA6r0hAGLwgKIC2sJECYXKDogUAsAOQ6gngiADhg949tu-a9LoCAhCg2BBEoaR4pA2DMG4PDGGUchrOo5p6BIuzmoc%2BxYP80wlroZavB814OreWFno%2BcDPq%2BRYOASArLIxGycSxbGcRxWwxBgZiQEQ3G8Vc6DloJGLaFpbQ3ph97iZJ%2BbSbJtLLDe4EoNMAAeeqUVJtiUXZEjLNBYx6f8Ehtq6SRdtMTlRAAjFEABMUQAMwLEAA)

Curious if there any specific reasons why this is the case? I suppose in most cases you would just want to return `Self` when creating a classmethod constructor, but that obviously doesn't preserve the generic param either.

Not sure if this has to do with any of the PEP 696 conformance changes.

**Comment by erictraut** [comment]
This change is due to a clarification in the typing spec that was mandated by [PEP 696](https://peps.python.org/pep-0696/).

When evaluating the expression `_BaseJob.from_callable`, the subexpression `_BaseJob` refers to a generic class that accepts a single type argument, but no type argument is provided here. PEP 696 adds the ability for the type parameter to have a default value. If no default value is provided, the type argument is `Any` (or `Unknown` in pyright to indicate that it's an "implied" `Any`). In your example, `_BaseJob.from_callable` evaluates to `(fn: Callable[..., Unknown], *args, Any, **kwargs: Any) -> _BaseJob[Unknown]`.

If you don't want the type argument to take on its default value of `Unknown`, you need to explicitly specify it when calling the class method.

```python
_BaseJob[int].from_callable(foo, 1, 2, 3)
```

**Reply by max-muoto** [reply]
Sounds good, I guess I should do this and utilize `Self` for the return-type, thanks for the clarification!
